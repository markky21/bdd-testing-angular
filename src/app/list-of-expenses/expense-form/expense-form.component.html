<section class="Wrapper">
  <h4>Add new expense</h4>

  <form class="Form" [formGroup]="form" (ngSubmit)="onSubmit(form)">
    <mat-form-field appearance="fill">
      <mat-label>Date of expense</mat-label>
      <input formControlName="date" matInput [matDatepicker]="picker" [attr.test-id]="'inputDate'" />
      <mat-datepicker-toggle matSuffix [for]="picker"></mat-datepicker-toggle>
      <mat-datepicker #picker></mat-datepicker>
    </mat-form-field>

    <mat-form-field appearance="fill">
      <mat-label>Amount</mat-label>
      <input formControlName="amount" matInput type="number" value="Sushi" [attr.test-id]="'inputAmount'" />
    </mat-form-field>

    <mat-form-field appearance="fill">
      <mat-label>Expense category</mat-label>
      <mat-select formControlName="category" [attr.test-id]="'inputCategory'">
        <mat-option *ngFor="let category of expenseCategory" [value]="expenseCategory[0]">
          {{ category }}
        </mat-option>
      </mat-select>
    </mat-form-field>

    <button type="submit" mat-raised-button color="primary" [attr.test-id]="'buttonSubmit'">Add expense</button>
  </form>
</section>
