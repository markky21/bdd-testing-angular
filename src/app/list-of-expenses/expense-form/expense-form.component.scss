.Wrapper {
  //display: block;
}
.Form {
  display: flex;
  width: 100%;
  align-items: center;

  > * {
    margin-right: 24px;
  }
}
