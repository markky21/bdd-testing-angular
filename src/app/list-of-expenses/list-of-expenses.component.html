<mat-card class="Wrapper">
  <mat-card-header>
    <mat-card-title>List of expenses</mat-card-title>
  </mat-card-header>
  <mat-card-content>
    <app-month-selector></app-month-selector>
    <app-expense-table></app-expense-table>
    <app-expense-form></app-expense-form>
  </mat-card-content>
</mat-card>
