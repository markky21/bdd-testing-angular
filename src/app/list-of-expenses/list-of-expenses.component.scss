.Form {
  margin: 32px;
}
