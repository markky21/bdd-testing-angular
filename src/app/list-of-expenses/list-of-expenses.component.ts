import { Component } from '@angular/core';

@Component({
  selector: 'app-list-of-expenses',
  templateUrl: './list-of-expenses.component.html',
  styleUrls: ['./list-of-expenses.component.scss'],
})
export class ListOfExpensesComponent {}
