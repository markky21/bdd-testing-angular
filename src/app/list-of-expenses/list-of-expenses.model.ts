export enum ExpenseCategory {
  HOME = 'HOME',
  CAR = 'CAR',
}

export interface Expense {
  id?: number;
  date: string;
  amount: number;
  category: ExpenseCategory;
}
