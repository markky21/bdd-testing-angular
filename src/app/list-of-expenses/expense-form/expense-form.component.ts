import { Component } from '@angular/core';
import { ExpenseCategory } from '../list-of-expenses.model';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { ListOfExpensesService } from '../list-of-expenses.service';

@Component({
  selector: 'app-expense-form',
  templateUrl: './expense-form.component.html',
  styleUrls: ['./expense-form.component.scss'],
})
export class ExpenseFormComponent {
  expenseCategory = Object.values(ExpenseCategory);
  form: FormGroup = this.formBuilder.group({
    date: ['', Validators.required],
    amount: ['', Validators.required],
    category: ['', Validators.required],
  });

  constructor(private formBuilder: FormBuilder, private listOfExpensesService: ListOfExpensesService) {}

  onSubmit(form: FormGroup): void {
    this.listOfExpensesService.dbAdd$(form.value).subscribe();
  }
}
