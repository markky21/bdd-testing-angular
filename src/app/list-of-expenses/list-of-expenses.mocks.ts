import * as moment from 'moment';
import { Expense, ExpenseCategory } from './list-of-expenses.model';

export const mockExpenseList: Expense[] = [
  {
    id: 1,
    date: moment().add(-5, 'days').toISOString(),
    amount: 15.5,
    category: ExpenseCategory.CAR,
  },
  {
    id: 2,
    date: moment().add(-9, 'days').toISOString(),
    amount: 163.5,
    category: ExpenseCategory.HOME,
  },
  {
    id: 3,
    date: moment().add(-10, 'days').toISOString(),
    amount: 5.5,
    category: ExpenseCategory.HOME,
  },
];

export const getMockExpenseList: () => Expense[] = () => [
  {
    id: 1,
    date: moment().add(-5, 'days').toISOString(),
    amount: 15.5,
    category: ExpenseCategory.CAR,
  },
  {
    id: 2,
    date: moment().add(-9, 'days').toISOString(),
    amount: 163.5,
    category: ExpenseCategory.HOME,
  },
  {
    id: 3,
    date: moment().add(-10, 'days').toISOString(),
    amount: 5.5,
    category: ExpenseCategory.HOME,
  },
];
