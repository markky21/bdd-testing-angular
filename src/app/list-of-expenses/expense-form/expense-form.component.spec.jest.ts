import { ComponentFixture, TestBed } from '@angular/core/testing';
import moment from 'moment';
import { ExpenseFormComponent } from './expense-form.component';
import { MatInputModule } from '@angular/material/input';
import { MatDatepickerModule } from '@angular/material/datepicker';
import { MatSelectModule } from '@angular/material/select';
import { MatNativeDateModule } from '@angular/material/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { ReactiveFormsModule } from '@angular/forms';
import { Overlay, OverlayContainer } from '@angular/cdk/overlay';
import { ListOfExpensesService } from '../list-of-expenses.service';
import { of } from 'rxjs';
import { ExpenseCategory } from '../list-of-expenses.model';
import { ListOfExpensesServiceStub } from '../list-of-expenses.service.stub';

// NOTE: In Angular Component tests we should mainly focus on testing component DOM structure changes
// and less on how things are implemented.

describe('ExpenseFormComponent', () => {
  let component: ExpenseFormComponent;
  let fixture: ComponentFixture<ExpenseFormComponent>;
  let service: ListOfExpensesService;
  let overlay: Overlay;
  let overlayContainer: OverlayContainer;

  const getTemplate = (): HTMLElement => fixture.debugElement.nativeElement;
  const getInputDateSelector = () =>
    getTemplate().querySelector('input[test-id="inputDate"]') as HTMLInputElement | null;
  const getInputAmountSelector = () =>
    getTemplate().querySelector('input[test-id="inputAmount"]') as HTMLInputElement | null;
  const getInputCategorySelector = () => getTemplate().querySelector('[test-id="inputCategory"]') as HTMLElement | null;
  const getSubmitButton = () => getTemplate().querySelector('[test-id="buttonSubmit"]') as HTMLButtonElement | null;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [ExpenseFormComponent],
      providers: [{ provide: ListOfExpensesService, useClass: ListOfExpensesServiceStub }],
      imports: [
        MatInputModule,
        MatDatepickerModule,
        MatSelectModule,
        MatNativeDateModule,
        NoopAnimationsModule,
        ReactiveFormsModule,
      ],
    }).compileComponents();

    service = TestBed.inject(ListOfExpensesService);
    overlayContainer = TestBed.inject(OverlayContainer);
    overlay = TestBed.inject(Overlay);

    fixture = TestBed.createComponent(ExpenseFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should display page components', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should user be able to add new expense', () => {
    const spyDbAdd = spyOn(service, 'dbAdd$').and.callFake(() => {
      return of([]);
    });
    // Set Date
    // tslint:disable-next-line:no-non-null-assertion
    getInputDateSelector()!.value = moment().format('M/DD/YYYY');
    getInputAmountSelector()?.dispatchEvent(new Event('input'));

    // Set Amount
    // tslint:disable-next-line:no-non-null-assertion
    getInputAmountSelector()!.value = '123';
    getInputAmountSelector()?.dispatchEvent(new Event('input'));

    // Select Category
    getInputCategorySelector()?.querySelector('.mat-select-trigger')?.dispatchEvent(new MouseEvent('click'));
    fixture.detectChanges();
    overlayContainer.getContainerElement().querySelector('.mat-option')?.dispatchEvent(new MouseEvent('click'));
    fixture.detectChanges();

    // Submit form
    getSubmitButton()?.click();
    fixture.detectChanges();
    expect(spyDbAdd).toHaveBeenCalledWith({ date: '', amount: 123, category: ExpenseCategory.HOME });
  });
});
