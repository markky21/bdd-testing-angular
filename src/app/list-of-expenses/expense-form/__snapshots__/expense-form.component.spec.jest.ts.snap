// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseFormComponent should display page components 1`] = `
<app-expense-form
  expenseCategory={[Function Array]}
  form={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  listOfExpensesService={[Function ListOfExpensesServiceStub]}
>
  <section
    class="Wrapper"
  >
    <h4>
      Add new expense
    </h4>
    <form
      class="Form ng-untouched ng-pristine ng-invalid"
      novalidate=""
    >
      <mat-form-field
        appearance="fill"
        class="mat-form-field mat-primary mat-form-field-type-mat-input mat-form-field-appearance-fill mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid _mat-animation-noopable"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-haspopup="dialog"
                aria-invalid="false"
                aria-required="false"
                class="mat-datepicker-input mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-invalid cdk-text-field-autofill-monitored"
                data-mat-calendar="mat-datepicker-0"
                formcontrolname="date"
                matinput=""
                test-id="inputDate"
              />
              <mat-datepicker />
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  class="mat-form-field-label mat-empty mat-form-field-empty"
                >
                  
                  
                  <mat-label>
                    Date of expense
                  </mat-label>
                  
                  
                </label>
              </span>
            </div>
            
            <div
              class="mat-form-field-suffix"
            >
              <mat-datepicker-toggle
                class="mat-datepicker-toggle"
                data-mat-calendar="mat-datepicker-0"
                matsuffix=""
              >
                <button
                  aria-haspopup="dialog"
                  aria-label="Open calendar"
                  class="mat-focus-indicator mat-icon-button mat-button-base _mat-animation-noopable"
                  mat-icon-button=""
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="mat-button-wrapper"
                  >
                    
                    <svg
                      class="mat-datepicker-toggle-default-icon"
                      fill="currentColor"
                      focusable="false"
                      height="24px"
                      viewBox="0 0 24 24"
                      width="24px"
                    >
                      <path
                        d="M19 3h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V8h14v11zM7 10h5v5H7z"
                      />
                    </svg>
                  </span>
                  <span
                    class="mat-button-ripple mat-ripple mat-button-ripple-round"
                    matripple=""
                  />
                  <span
                    class="mat-button-focus-overlay"
                  />
                </button>
              </mat-datepicker-toggle>
            </div>
          </div>
          
          <div
            class="mat-form-field-underline"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
          >
            
            
            <div
              class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
      <mat-form-field
        appearance="fill"
        class="mat-form-field mat-primary mat-form-field-type-mat-input mat-form-field-appearance-fill mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid _mat-animation-noopable"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-invalid="false"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-invalid cdk-text-field-autofill-monitored"
                formcontrolname="amount"
                matinput=""
                test-id="inputAmount"
                type="number"
                value="Sushi"
              />
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  class="mat-form-field-label mat-empty mat-form-field-empty"
                >
                  
                  
                  <mat-label>
                    Amount
                  </mat-label>
                  
                  
                </label>
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-underline"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
          >
            
            
            <div
              class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
      <mat-form-field
        appearance="fill"
        class="mat-form-field mat-primary mat-form-field-type-mat-select mat-form-field-appearance-fill mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid _mat-animation-noopable"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <mat-select
                aria-autocomplete="none"
                aria-disabled="false"
                aria-expanded="false"
                aria-haspopup="true"
                aria-invalid="false"
                aria-required="false"
                class="mat-select mat-select-empty ng-untouched ng-pristine ng-invalid"
                formcontrolname="category"
                role="combobox"
                tabindex="0"
                test-id="inputCategory"
              >
                <div
                  cdk-overlay-origin=""
                  class="mat-select-trigger"
                >
                  <div
                    class="mat-select-value"
                  >
                    
                    <span
                      class="mat-select-placeholder mat-select-min-line"
                    >
                      
                    </span>
                    
                  </div>
                  <div
                    class="mat-select-arrow-wrapper"
                  >
                    <div
                      class="mat-select-arrow"
                    />
                  </div>
                </div>
                
              </mat-select>
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  class="mat-form-field-label mat-empty mat-form-field-empty"
                >
                  
                  
                  <mat-label>
                    Expense category
                  </mat-label>
                  
                  
                </label>
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-underline"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
          >
            
            
            <div
              class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
      <button
        color="primary"
        mat-raised-button=""
        test-id="buttonSubmit"
        type="submit"
      >
        Add expense
      </button>
    </form>
  </section>
</app-expense-form>
`;
